### PR TITLE
Merge configmap and secret values

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#4949

This PR implements the merging of configmap and secrets using `helmclient.MergeValues`.

The values we get back from the Tiller API are not identical to the ones we provide. So an annotation 
`chart-operator.giantswarm.io/values-md5-checksum` is used for the comparison in the reconciliation logic.

### CODE REVIEW

File: service/controller/chart/v1/resource/release/desired.go [hunk 1]
```diff
@@ -40,41 +42,59 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
...
+		valuesYAML, err = yaml.Marshal(values)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		valuesMD5Checksum = fmt.Sprintf("%x", md5.Sum(valuesYAML))
+	} else {
+		// We need to pass empty values in ValueOverrides to make the install
+		// process use the default values and prevent errors on nested values.
+		valuesYAML = []byte("{}")
```
Comment (rossf7) [inline service/controller/chart/v1/resource/release/desired.go]: We still need to pass empty values if no values have been specified. This is so Helm uses the `values.yaml` for the chart.

File: service/controller/chart/v1/resource/release/desired_test.go [hunk 2]
```diff
@@ -343,93 +348,20 @@ func Test_DesiredState(t *testing.T) {
 				t.Fatalf("error == %#v, want matching", err)
 			}
 
-			ReleaseState, err := toReleaseState(result)
+			releaseState, err := toReleaseState(result)
```
Comment (rossf7) [inline service/controller/chart/v1/resource/release/desired_test.go]: Naming was wrong so sneaking in this fix.

File: service/controller/chart/v1/resource/release/resource.go [hunk 3]
```diff
@@ -89,6 +93,39 @@ func (r *Resource) Name() string {
 	return Name
 }
 
+func (r *Resource) updateAnnotations(ctx context.Context, cr v1alpha1.Chart, releaseState ReleaseState) error {
```
Comment (tomahawk28) [inline service/controller/chart/v1/resource/release/resource.go]: This function do 3 different things. 
1. Get the current Chart. 
2. Compare md5sum
3. Update `Chart` with a new one. 
But just name it `updateAnnotation` is a bit unclear for me. 
Can we divide the function of inside, or simply rename it?

Comment (rossf7) [reply service/controller/chart/v1/resource/release/resource.go]: The comparison is so we only make a k8s API call if the annotation has changed. The reconciliation logic uses `isReleaseModified`. Should I add more comments to make that clearer?

File: service/controller/chart/v1/resource/release/resource.go [hunk 4]
```diff
@@ -89,6 +93,39 @@ func (r *Resource) Name() string {
...
+		annotations[key.ValuesMD5ChecksumAnnotationName] = releaseState.ValuesMD5Checksum
+
+		currentCR.Annotations = annotations
+		_, err = r.g8sClient.ApplicationV1alpha1().Charts(cr.Namespace).Update(currentCR)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated annotations for app CR %#q in namespace %#q", cr.Name, cr.Namespace))
```
Comment (tomahawk28) [inline service/controller/chart/v1/resource/release/resource.go]: This log only related to when we actually update `annotation`

File: service/controller/chart/v1/resource/release/resource.go (hunk 3 above)
Comment (tomahawk28) [reply service/controller/chart/v1/resource/release/resource.go]: Maybe some `godoc` for the function is enough then.

File: service/controller/chart/v1/resource/release/resource.go (hunk 4 above)
Comment (rossf7) [reply service/controller/chart/v1/resource/release/resource.go]: Good point. I'll fix the logging and add the missing godoc. Thx!

**Comment by rossf7** [comment]
@tomahawk28 PTAL

**Pull request merged**